### ISSUE TITLE
Button support

### ISSUE THREAD

**Issue by Elvys8** [mdaskalov/homebridge-tasmota-zbbridge#15]
Hi, sorry if what I'm asking is silly or obvious…
I've just flashed my ZB-GW03 and paired a couple of devices. Lights work as expected, but I see no way of making buttons appear on the Home App. I've paired an Aqara Wireless Switch D1 and an IKEA Trådfri shortcut button. Both show up in the Tasmota web interface correctly, but when exposed to HomeKit they appear as switches, not buttons, so I can't asign scenes to them.
Am I doing something wrong?

Thanks a lot!

**Plugin Config:**

```json
{
            "name": "Tasmota ZbBridge",
            "mqttTopic": "tasmota_D2C9C0",
            "zbBridgeDevices": [
                {
                    "addr": "0xB8ED",
                    "type": "light0",
                    "name": "Neón Cocina"
                },
                {
                    "addr": "0xAACF",
                    "type": "switch",
                    "name": "Dormitorio Interruptor Entrada"
                },
                {
                    "addr": "0xDA5C",
                    "type": "switch",
                    "name": "IKEA1"
                }
            ],
            "mqttBroker": "192.168.1.100:1883",
            "platform": "TasmotaZbBridge"
        }
```

**Environment:**

* **Plugin Version**: v1.1.0
* **Homebridge Version**: v1.4.1
* **Node.js Version**: v16.15.0
* **NPM Version**: v8.5.5
* **Operating System**:  Docker

**Comment by mdaskalov** [comment]
I was wandering lately about the same...
So far as I know there are no buttons in Homebridge (maybe HomeKit also). According to the [documentation](https://developers.homebridge.io/#/service/Doorbel) the closest service would be `Doorbell` supporting single, double and long press of the button. You can try [homebridge-mqttthing](https://www.npmjs.com/package/homebridge-mqttthing) plugin and if it’s working as expected I can also implement this functionality in my plugin.

**Comment by mdaskalov** [comment]
Well, after some research it turned out, that there is indeed a button functionality - the [StatelessProgrammableSwitch](https://developers.homebridge.io/#/service/StatelessProgrammableSwitch) which we can use with the `sensor` accessory.

I’ve implemented it as generic accessory which can cover all service types (read-only).
To configure we need to know the correct cluster and attribute which the button is sending. The configuration should look as follows:

```
{
    "addr": "0xDA5C",
    "type": "sensor_<cluster>_<attribute>_StatelessProgrammableSwitch_ProgrammableSwitchEvent_<event>",
    "name": "ButtonSensor"
}
```

Please quote here the `ZbReceived` message (for single, double and long press) from the console so we can extract the correct values for `cluster`, `attribute` and `event` above.

**Comment by Elvys8** [comment]
No sure if I'm doing this right, but let's go…

**Single press:**
`07:39:17.883 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0xDA5C":{"Device":"0xDA5C","Name":"IKEA1","0006!01":"","Power":1,"Endpoint":1,"LinkQuality":68}}}`

**Double press:**
(same result, maybe not available in this button)
`07:39:20.284 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0xDA5C":{"Device":"0xDA5C","Name":"IKEA1","0006!01":"","Power":1,"Endpoint":1,"LinkQuality":81}}}`

**Long press:**
`07:39:23.033 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0xDA5C":{"Device":"0xDA5C","Name":"IKEA1","0008!05":"0053","DimmerMove":0,"Endpoint":1,"LinkQuality":84}}}
07:39:24.833 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0xDA5C":{"Device":"0xDA5C","Name":"IKEA1","0008!07":"","DimmerStop":true,"Endpoint":1,"LinkQuality":84}}}
`

**Comment by Elvys8** [comment]
Log from another button, in case it helps.
This is from a [Xiaomi Mi Smart Wireless Switch](https://mistore-dphone.com/xiaomi-mi-smart-wireless-switch).

**Single press:**

```
13:00:06.618 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0x3077":{"Device":"0x3077","Power":0,"Endpoint":1,"LinkQuality":84}}}
13:00:06.992 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0x3077":{"Device":"0x3077","Power":1,"Endpoint":1,"LinkQuality":84}}}
```
**Double press:**
```
13:00:36.002 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0x3077":{"Device":"0x3077","Power":2,"Endpoint":1,"LinkQuality":76}}}
```


**Long press:**
(same as single press, apparently)
```
13:00:52.644 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0x3077":{"Device":"0x3077","Power":0,"Endpoint":1,"LinkQuality":97}}}
13:00:54.152 MQT: tele/tasmota_D2C9C0/SENSOR = {"ZbReceived":{"0x3077":{"Device":"0x3077","Power":1,"Endpoint":1,"LinkQuality":97}}}
```

**Comment by mdaskalov** [comment]
From what I can see both are switches and not buttons. You can configure them as button using cluster 6, attribute 1 and event "Power" but it's easier to just configure them as a switch.

I have my IKEA switch bound to all lamps directly (not visible in HomeKit) so I can control dimming (holding on/off) and switching on/off at the same time. It also happens directly - not trough HomeKit.

See the example in the README file.

**Comment by Elvys8** [comment]
Your setup is perfect for an all zigbee ecosystem, but right now I'm controlling a lot of devices, wifi fans, bluetooth locks, Smart TV and speakers… It's much more convenient for me to have exposed HomeKit buttons I can assign scenes and shortcuts to from my iPhone.

I'll try to configure them as buttons, but not sure my "coding" skills are up to the task 😅

**Comment by mdaskalov** [comment]
The [StatelessProgrammableSwitch](https://developers.homebridge.io/#/service/StatelessProgrammableSwitch) can recognize these events (see [here](https://developers.homebridge.io/#/characteristic/ProgrammableSwitchEvent)):
```
SINGLE_PRESS  0
DOUBLE_PRESS  1
LONG_PRESS    2
```
So if we use the `Power` value your Mi Switch will fire `SINGLE_PRESS` then `DOUBLE_PRESS` on single press of the button. It will fire `LONG_PRESS` on double press of the button.

The configuration should look like follows (The Power command belongs to cluster 6 attribute 1, see explanations [here](https://tasmota.github.io/docs/Zigbee/#understanding-endpoints-and-clusters)):
```
{
    "addr": "0x3077",
    "type": "sensor_6_1_StatelessProgrammableSwitch_ProgrammableSwitchEvent_Power",
    "name": "MiButton"
}
```
For the IKEA button we can’t configure single HomeKit button for short and long press, as they use different clusters and values. 
The configuration for single press is as follows (will fire `DOUBLE_PRESS` using the `Power` value):
```
{
    "addr": "0xDA5C",
    "type": "sensor_6_1_StatelessProgrammableSwitch_ProgrammableSwitchEvent_Power",
    "name": "IKEA_Single"
}
```
The configuration for long press is as follows (will fire `SINGLE_PRESS` using the `DimmerMove` value):
```
{
    "addr": "0xDA5C",
    "type": "sensor_8_5_StatelessProgrammableSwitch_ProgrammableSwitchEvent_DimmerMove",
    "name": "IKEA_Single"
}
```

**Comment by Elvys8** [comment]
Wow! Thanks a lot for taking the time to break it down for me. I’m on holiday right now, I’ll give it a try as soon as I get home. 😊

**Comment by mdaskalov** [comment]
Resolved in #19
